### DISCUSSION TITLE
Binder link for Pluto notebook.

### DISCUSSION THREAD

**Discussion by curtcorum** [JuliaHealth/KomaMRI.jl#399] [Show and tell]
Here is a link that can generate a binder instance from a Pluto notebook:

https://pluto-on-binder.glitch.me/

I fed it the reproducibility notebook at:

https://raw.githubusercontent.com/JuliaHealth/KomaMRI.jl/master/examples/4.reproducible_notebooks/pluto-01-gradient-echo-spin-echo.jl

and generated the following:

https://binder.plutojl.org/v0.19.36/open?url=https%253A%252F%252Fraw.githubusercontent.com%252FJuliaHealth%252FKomaMRI.jl%252Fmaster%252Fexamples%252F4.reproducible_notebooks%252Fpluto-01-gradient-echo-spin-echo.jl

First time it ran for 1500 seconds and then stalled, in the precomplie step.

**Comment by curtcorum** [comment]
The local first run was pretty quick on my machine:

![local_first](https://github.com/JuliaHealth/KomaMRI.jl/assets/1796102/ddfe2b70-c992-45e7-8692-de689185232b)

**Comment by curtcorum** [comment]
Does anybody know of a better place to try Pluto notebooks (or Jupyter) online/cloud than mybinder.org?

**Comment by cncastillo** [comment]
Hi! I didn't know about [pluto-on-binder.glitch.me](https://pluto-on-binder.glitch.me/), but it seems to be doing the same as clicking the binder option in the notebook:

<table>
<tr>
 <td> <img src="https://github.com/JuliaHealth/KomaMRI.jl/assets/5957134/3ef5ada0-1668-44a6-9ae0-6a4417528642">
 <td> <img src="https://github.com/JuliaHealth/KomaMRI.jl/assets/5957134/2acd0fe2-f976-4a22-8eb6-149d5845e534">
 <td> <img src="https://github.com/JuliaHealth/KomaMRI.jl/assets/5957134/899f7659-449b-46fd-b64a-52d67741e5c9">
</table>

Yesterday, I tried both approaches (pluto-on-binder and running binder directly in the notebook), and they are quite slow, as they are compiling the whole environment first. I also tried https://cocalc.com/ with little success (another option is https://glassnotebook.io/, which I didn't try). 

# Jupyter notebooks on Binder

For Julia packages, MyBinder has a direct way of creating a docker image using the package GitHub repo. As this environment is already compiled, it just needs to download the image, which is a LOT faster and it works quite well for the **tutorials** (jupyter notebooks that work on Binder's JupyterLab instance). For example [this tutorial](https://juliahealth.org/KomaMRI.jl/dev/tutorial/01-FID/). 

Launch KomaMRI using Binder's pre-built docker image [![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/JuliaHealth/KomaMRI.jl/HEAD)

<table>
<tr>
 <td> <img src="https://github.com/JuliaHealth/KomaMRI.jl/assets/5957134/bcbd4e2a-251a-4184-9efe-fef134e15069">
 <td> <img src="https://github.com/JuliaHealth/KomaMRI.jl/assets/5957134/bd17bfe3-800d-4761-bf3a-5d070a8c3925">
<td> <img src="https://github.com/JuliaHealth/KomaMRI.jl/assets/5957134/f95e64af-b431-4f50-9ba2-1d7eb9f2cdca">
<tr> 
 <td> <img src="https://github.com/JuliaHealth/KomaMRI.jl/assets/5957134/e22025b5-4526-4f74-b826-0e2c28050bfc">
</table>

This approach had a problem with Plotly figures ([not WebIO found](https://discourse.jupyter.org/t/jupyter-webio-in-binder/14476)), but I added an `environment.yml` [yesterday](https://github.com/JuliaHealth/KomaMRI.jl/blob/b3d6cd4fcd915c618fc22e600eecfe639b0ba529/environment.yml) to configure Binder's Jupyter lab properly (as the [new JupyterLab does not work with WebIO](https://github.com/JuliaGizmos/WebIO.jl/issues/514)). Now that this works, I will add a Binder link to each tutorial.

# Potential solution

Using a slightly modified version of that docker image, we could potentially launch a Pluto server on Binder, by adding `jupyter-server-proxy` to `environment.yml` and implementing this: https://github.com/fonsp/pluto-on-binder/issues/21. But we would need to add `Pluto.jl` to `KomaMRI.jl`'s dependencies (?); maybe Pluto can be installed using a `Dockerfile` so we don't modify Koma's dependencies. 

Whatever approach we decide, it would be nice to add launch-on binder buttons to every tutorial (reproducible or not) :smile: , but I would need some help with the Pluto notebooks.

**Reply by curtcorum** [reply]
Carlos, thanks for looking into this. I agree that having notebooks that can go live for examples, tutorials, etc. are great things. I'll try your suggestions.

**Comment by cncastillo** [comment]
Potentially useful links for setting up a Koma+Pluto image: 

- [Mybinder.org broken for Julia repos?](https://discourse.julialang.org/t/mybinder-org-broken-for-julia-repos/108743)
- [Hosting pluto notebooks online - 2024](https://discourse.julialang.org/t/hosting-pluto-notebooks-online-2024/114920)
- Oceanigans: https://github.com/CliMA/Oceananigans.jl/blob/main/Dockerfile

**Comment by curtcorum** [comment]
# jupyter-pluto-proxy

This python package ads the capability to launch a Pluto notebook from JupyterLab and also has an example of jumping directly to Pluto. I am taking a look at various scenerios, but it could be added to KomaMRI so that the docker image has it installed.

I'm guessing the Pluto dependency for the docker image can be kept separate from the KomaMRI.jl dependencies

Launching from the KomaMRI prebuilt docker is definitely the way to go!

https://github.com/yuvipanda/jupyter-pluto-proxy

**Comment by curtcorum** [comment]
Examples from https://github.com/yuvipanda/jupyter-pluto-proxy

This launches directly into Pluto:
[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/yuvipanda/jupyter-pluto-proxy/main?urlpath=pluto)

This launches into JupyterLab, with option to open Pluto notebook:
[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/yuvipanda/jupyter-pluto-proxy/main)

![jupyterlab_launch_pluto](https://github.com/JuliaHealth/KomaMRI.jl/assets/1796102/76a64def-c625-4b1b-ad02-d14ee537deed)

**Comment by curtcorum** [comment]
# See PR https://github.com/JuliaHealth/KomaMRI.jl/pull/400 which should let it be tested with KomaMRI docker binder instance.

It just adds `jupyter-pluto-proxy` to `environment.yml`
No doubt some more tweaks will be needed, and this should not mess anything else up (famous last words?!).

**Comment by curtcorum** [comment]
https://github.com/JuliaHealth/KomaMRI.jl/pull/400

Still needed is a way to launch a Pluto notebook file along with the (KomaMRI) docker image into binder from a badge/url.

A secondary issue is to launch a pluto-notebook from JupyterLab without first launching the Pluto instance.

![pluto_binder_koma_notebook](https://github.com/JuliaHealth/KomaMRI.jl/assets/1796102/c9ac852b-644e-426e-9e88-2264f8fcbbe3)

**Comment by curtcorum** [comment]
# Sample Binder Repositories
Below we list several sample Binder repositories that demonstrate how to compose build files in order to createBinders with varying environments. You can find all of the repositories listed on this page at the binder-examples GitHub organization.

https://mybinder.readthedocs.io/en/latest/examples/sample_repos.html

**Comment by curtcorum** [comment]
# <a href="https://github.com/jupyterhub/repo2docker"><img src="https://raw.githubusercontent.com/jupyterhub/repo2docker/71eb8058c790a88d223470a55f3ea5b744614dcf/docs/source/_static/images/repo2docker.png" height="40px" /></a>  repo2docker GitHub Action

Trigger [repo2docker](https://github.com/jupyter/repo2docker) to build a Jupyter enabled Docker image from your GitHub repository and push this image to a Docker registry of your choice. This will automatically attempt to build an environment from configuration files found in your repository in the [manner described here](https://repo2docker.readthedocs.io/en/latest/usage.html#where-to-put-configuration-files).

https://github.com/jupyterhub/repo2docker-action

**Comment by curtcorum** [comment]
# Use GitHub Actions To Cache The Build For BinderHub

Instead of forcing mybinder.org to cache your builds, you can optionally build a Docker image with GitHub Actions and push that to a Docker registry, so that any [BinderHub](https://binderhub.readthedocs.io/en/latest/) instance, including mybinder.org only has to pull the image. This might give you more control than triggering a build directly on mybinder.org like the method illustrated above. In this example, you must supply the [secrets](https://help.github.com/en/actions/configuring-and-managing-workflows/creating-and-storing-encrypted-secrets) DOCKER_USERNAME and DOCKER_PASSWORD so that Actions can push to DockerHub. Note that, instead of your actual password, you can use an [access token](https://docs.docker.com/docker-hub/access-tokens/) — which may be a more secure option.

https://github.com/jupyterhub/repo2docker-action?tab=readme-ov-file#use-github-actions-to-cache-the-build-for-binderhub

**Comment by curtcorum** [comment]
# Links to Try (still testing more...)

@cncastillo 

## opens up default JupyterLab:
https://mybinder.org/v2/gh/curtcorum/KomaMRI.jl/docker_jupyterlab_pluto

## opens a specific notebook in JupyterLab
https://mybinder.org/v2/gh/curtcorum/KomaMRI.jl/docker_jupyterlab_pluto?labpath=binder_notebooks%2F01-FID.ipynb
https://mybinder.org/v2/gh/curtcorum/KomaMRI.jl/docker_jupyterlab_pluto?filepath=binder_notebooks/01-FID.ipynb
https://mybinder.org/v2/gh/curtcorum/KomaMRI.jl/docker_jupyterlab_pluto?urlpath=https%3A%2F%2Fraw.githubusercontent.com%2Fcurtcorum%2FKomaMRI.jl%2Fdocker_jupyterlab_pluto%2Fbinder_notebooks%2F01-FID.ipynb

## opens up Jupyter classic:


## opens up Pluto
https://mybinder.org/v2/gh/curtcorum/KomaMRI.jl/docker_jupyterlab_pluto?urlpath=pluto/



## also opens up JupyterLab:
https://mybinder.org/v2/gh/curtcorum/KomaMRI.jl/docker_jupyterlab_pluto?filepath=
https://mybinder.org/v2/gh/curtcorum/KomaMRI.jl/docker_jupyterlab_pluto?urlpath=
https://mybinder.org/v2/gh/curtcorum/KomaMRI.jl/docker_jupyterlab_pluto?urlpath=/

## also opens up Jupyter classic:
https://mybinder.org/v2/gh/curtcorum/KomaMRI.jl/docker_jupyterlab_pluto?filepath=/

**Comment by cncastillo** [comment]
To open Pluto notebooks directly, there is an example here 
https://github.com/roualdes/plutonotebooks
https://mybinder.org/v2/gh/roualdes/plutonotebooks/main?urlpath=pluto/open?path=/home/jovyan/notebooks/beta.jl

you just need to add `open?path=notebook_path_in_binder.jl`

**Comment by curtcorum** [comment]
Carlos, 

Awesome, I will try it out!

Curt

On 5 June 2024 3:49:47 am UTC, Carlos Castillo Passi ***@***.***> wrote:
>To open Pluto notebooks directly, there is an example here 
>https://github.com/roualdes/plutonotebooks
>https://mybinder.org/v2/gh/roualdes/plutonotebooks/main?urlpath=pluto/open?path=/home/jovyan/notebooks/beta.jl
>
>you just need to add `open?path=notebook_path_in_binder.jl`
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/JuliaHealth/KomaMRI.jl/discussions/399#discussioncomment-9672413
>You are receiving this because you authored the thread.
>
>Message ID: ***@***.***>
-- 
Curtis A. Corum, Ph.D.
President and CEO
Champaign Imaging LLC

***@***.***
www.champaignimaging.com

Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by cncastillo** [comment]
Hi! I managed to fix this in `dev`!  Now a Binder button should appear in all the tutorials :)

https://juliahealth.org/KomaMRI.jl/dev/
![image](https://github.com/JuliaHealth/KomaMRI.jl/assets/5957134/5436bb51-d2d0-442a-9ebc-59224fc4dcf2)

- Literate tutorial: https://mybinder.org/v2/gh/JuliaHealth/KomaMRI.jl/master?urlpath=git-pull%3Frepo%3Dhttps%253A%252F%252Fgithub.com%252FJuliaHealth%252FKomaMRI.jl%26urlpath%3Dlab%252Ftree%252FKomaMRI.jl%252Fdev%252Ftutorial%252F01-FID.ipynb%26branch%3Dgh-pages
- Pluto tutorial: https://mybinder.org/v2/gh/JuliaHealth/KomaMRI.jl/master?urlpath=git-pull%3Frepo%3Dhttps%253A%252F%252Fgithub.com%252FJuliaHealth%252FKomaMRI.jl%26urlpath%3Dpluto%252Fopen%3Fpath%3DKomaMRI.jl%252Fdev%252Ftutorial-pluto%252F01-gradient-echo-spin-echo.jl%26branch%3Dgh-pages

Sometimes, building the docker image times out, but that will be fixed when we tag v0.9, which removes CUDA as a strong dependency. 

To make Pluto notebooks run faster modify the package-loading cell to 
```julia
begin
	import Pkg
	Pkg.activate("/home/jovyan/.binder/")
	using KomaMRI, PlutoPlotly, PlutoUI # or whatever was included
end
```

Could you mark this as the solution?
